### DISCUSSION TITLE
Externalise Web Part Settings into a Config File

### DISCUSSION THREAD

**Discussion by 5ur3nP** [microsoft-search/pnp-modern-search#2771] [Ideas]
**Is your feature request related to a problem? Please describe.**
I have used PnP Search Results on many site pages to query and display related items to the page itself.  However, it is difficult to add new features since one would have to update the configuration of the web part on each of several hundred items - somewhat impractical.

**Describe the solution you'd like**
I would like the web part to be able to load its settings from a config file that is stored somewhere in the Sharepoint site.  This means that there is a single place to change the config of all of the different instances of Search Results across many pages.  This  could potentially be done in a similar way to the Import/Export settings capability, which produces a JSON file.  This could be similar to the way that "Use an external template URL" is implemented.

**Describe alternatives you've considered**
It may be possible to develop custom web parts that consume configuration files, but this would not be as feature rich or capable as PnP search.

**Comment by kasperbolarsen** [comment]
There are a number of ways to solve this issue.  Extracting the updated web part using pnp provisioning and apply this to the sites/pages could be one option.  
The one I prefer is using this script: https://pnp.github.io/script-samples/spo-deploy-pnpmodernsearch-webpart/README.html?tabs=pnpps

**Reply by 5ur3n** [reply]
Thank you Kasper. I do think that the config externalisation would allow update of all web parts without needing an administrator, but this could work too. 'll touch base with the team on my end.

**Reply by 5ur3n** [reply]
We have several instances of search results on a page template, each having different configs. This template is used to multiple pages (KnowledgeBase articles which cross connect to delivery roadmaps, etc) The challenge is that the PNP provisioning script will not be able to tell them apart and the team has no way of filtering them. Hence we still think that externalisation of the config would be helpful.

**Reply by kasperbolarsen** [reply]
Hi @5ur3nP I think the best approach for you would be to create a fork of the PnP Modern Search and adapt it to fit your need. As externalisation of the configuration would creating a lot of complexity and making adding any new feature in the PnP Modern Search even harder to implement, I think it is unlikely that any PR to this effect will be accepted.
Personally, I would take a second look at the scripting option as a short (or permanent) solution.
